homeassistant/components/qnap/sensor.py

```diff
@@ -118,179 +182,90 @@
-    if not coordinator.last_update_success:
-        raise PlatformNotReady
-
-    monitored_conditions = config[CONF_MONITORED_CONDITIONS]
-    sensors: list[QNAPSensor] = []
-
-    # Basic sensors
-    sensors.extend(
-        [
-            QNAPSystemSensor(coordinator, description)
-            for description in _SYSTEM_MON_COND
-            if description.key in monitored_conditions
-        ]
-    )
-    sensors.extend(
-        [
-            QNAPCPUSensor(coordinator, description)
-            for description in _CPU_MON_COND
-            if description.key in monitored_conditions
-        ]
-    )
-    sensors.extend(
-        [
-            QNAPMemorySensor(coordinator, description)
-            for description in _MEMORY_MON_COND
-            if description.key in monitored_conditions
-        ]
-    )
-
-    # Network sensors
-    sensors.extend(
-        [
-            QNAPNetworkSensor(coordinator, description, nic)
-            for nic in config.get(CONF_NICS, coordinator.data["system_stats"]["nics"])
-            for description in _NETWORK_MON_COND
-            if description.key in monitored_conditions
-        ]
-    )
-
-    # Drive sensors
-    sensors.extend(
-        [
-            QNAPDriveSensor(coordinator, description, drive)
-            for drive in config.get(CONF_DRIVES, coordinator.data["smart_drive_health"])
-            for description in _DRIVE_MON_COND
-            if description.key in monitored_conditions
-        ]
-    )
-
-    # Volume sensors
-    sensors.extend(
-        [
-            QNAPVolumeSensor(coordinator, description, volume)
-            for volume in config.get(CONF_VOLUMES, coordinator.data["volumes"])
-            for description in _VOLUME_MON_COND
-            if description.key in monitored_conditions
-        ]
-    )
-
-    add_entities(sensors)
-
-
-def round_nicely(number):
-    """Round a number based on its size (so it looks nice)."""
-    if number < 10:
-        return round(number, 2)
-    if number < 100:
-        return round(number, 1)
-
-    return round(number)
-
 
 class QNAPSensor(CoordinatorEntity[QnapCoordinator], SensorEntity):
     """Base class for a QNAP sensor."""
 
     def __init__(
         self,
         coordinator: QnapCoordinator,
-        description: SensorEntityDescription,
-        monitor_device: str | None = None,
+        description,
+        uid,
+        monitor_device=None,
+        monitor_subdevice=None,
     ) -> None:
         """Initialize the sensor."""
-        super().__init__(coordinator)
+        self.coordinator = coordinator
         self.entity_description = description
-        self.monitor_device = monitor_device
+        self.uid = uid
         self.device_name = self.coordinator.data["system_stats"]["system"]["name"]
+        self.monitor_device = monitor_device
+        self.monitor_subdevice = monitor_subdevice
+
+    @property
+    def unique_id(self):
+        """Return unique_id."""
+        return f"{self.uid}_{self.name}"
```

Comment:
Please drop the property and use `self._attr_unique_id` instead.
Also, your unique_id is not so unique! Two different nics/drives/volumes would end up having the same unique ids.

You should also remove the name from the unique id - and use instead the description key which will be more robust. Maybe something like `entry_id + description_key + device_id (nic_id/drive/volume_id)`